File: apps/spell-circle-canvas/src/common/compose/reference/examples/text_element.py
"""text — the content forms: the inheriting leaf, the leaf set in a whole
style, and mixed text as a comparable value.

The Python twin of text_element.cpp. A reference example: it is rendered
with ``sigil render`` and belongs to no sketch registry.
"""

from sigil import compose, weave
from sigil.sketch import SketchContext, sketch

GROUND = "#14181d"
INK = "#e2e9ee"
ASH = "#8ea0ad"
ACCENT = "#e0a03c"


def stated() -> weave.TextStyle:
    """A total style: a ``weave.TextStyle`` states every field itself, so
    a leaf set in one inherits nothing from the tree above it."""
    return weave.textStyle(weave.Type(size=26, color=ACCENT))


def row(caption: str, leaf: compose.Element) -> compose.Element:
    return compose.box(compose.text(caption, size=12, color=ASH), leaf).column().gap(6)


@sketch(size=(620, 300), background=GROUND, capture_at=0)
class TextElement:
    def setup(self, ctx: SketchContext) -> None:
        ctx.render(self.describe())

    def describe(self) -> compose.Element:
        return (
            compose.box(
                row("text(utf8)", compose.text("Set in the font and ink in force.")),
                row(
                    "text(utf8, style)",
                    compose.text("Set in a style of its own.", stated()),
                ),
                row(
                    "text(rich)",
                    compose.text(
                        weave.rich()
                        .add("Mixed text as ")
                        .add("one comparable value", weave.Type(color=ACCENT))
                    ),
                ),
            )
            .column()
            .gap(26)
            .padding(28)
            # The font and the ink everything under this node is set in.
            .fontSize(20)
            .ink(INK)
        )

File: apps/spell-circle-canvas/src/common/compose/reference/examples/text_element.cpp
/** @file
 * text — the content forms: the inheriting leaf, the leaf set in a whole
 * style, and mixed text as a comparable value.
 *
 * A reference example. It stands outside the sketch registry, so it is
 * photographed with `--frame` and never enters the plate sweep.
 */

#include <sigilcompose/core/Core.h>
#include <sigilmaterial/color/Color.h>
#include <sigilsketch/canvas/Sketch.h>
#include <sigilweave/paragraph/RichText.h>
#include <sigilweave/style/Style.h>

namespace material = sigil::material;
namespace sketch = sigil::sketch;
namespace weave = sigil::weave;

using namespace sigil::compose;

namespace {

constexpr SkSize kCanvas = {620, 300};
constexpr material::Color kGround = hexColor(0x14181d);
constexpr material::Color kInk = hexColor(0xe2e9ee);
constexpr material::Color kAsh = hexColor(0x8ea0ad);
constexpr material::Color kAccent = hexColor(0xe0a03c);

/** A total style: a `weave::TextStyle` states every field itself, so a
 *  leaf set in one inherits nothing from the tree above it. */
weave::TextStyle stated() {
  weave::TextStyle style;
  style.shaping.fontSize = 26;
  style.paint.foreground.setColor4f(kAccent, nullptr);
  return style;
}

Element row(const char* caption, Element leaf) {
  return box().column().gap(6).children(
      {text(caption).font({.size = 12, .color = kAsh}), std::move(leaf)});
}

}  // namespace

struct TextElement {
  void setup(sketch::SketchContext& ctx) {
    ctx.canvas({.size = kCanvas, .background = kGround, .captureSeconds = 0});
    ctx.composer.render(describe());
  }

  Element describe() const {
    return box()
        .column()
        .gap(26)
        .padding(28)
        // The font and the ink everything under this node is set in.
        .font({.size = 20})
        .ink(kInk)
        .children({
            row("text(utf8)", text("Set in the font and ink in force.")),
            row("text(utf8, style)",
                text("Set in a style of its own.", stated())),
            row("text(rich)", text(weave::rich()
                                       .add("Mixed text as ")
                                       .add("one comparable value",
                                            weave::Type{.color = kAccent}))),
        });
  }
};

SIGIL_SKETCH(TextElement, "Reference · Compose",
             "the text leaf in its three content forms")
